pyccel/ast/cudaext.py

```diff
@@ -0,0 +1,309 @@
+            dim = LiteralInteger(dim)
+        if not isinstance(dim, LiteralInteger):
+            raise TypeError("dimension need to be an integer")
+        if dim not in (0, 1, 2):
+            raise ValueError("dimension need to be 0, 1 or 2")
+        #...
+        self._dim       = dim
+        self._shape     = None
+        self._rank      = 0
+        self._dtype     = dim.dtype
+        self._precision = dim.precision
+        self._order     = None
+        super().__init__()
+
+    @property
+    def dim(self):
+        return self._dim
+
+
+class CudaCopy(CudaNewArray):
+    """
+    Represents a call to  cuda.copy for code generation.
+
+    Parameters
+    ----------
+    arg : Variable
+
+    memory_location : str
+        'host'   the newly created array is allocated on host.
+        'device' the newly created array is allocated on device.
+    
+    is_async: bool
+        Indicates whether the copy is asynchronous or not [Default value: False]
+
+    """
+    __slots__ = ('_arg','_dtype','_precision','_shape','_rank','_order','_memory_location', '_is_async')
+
+    def __init__(self, arg, memory_location, is_async=False):
+        
+        if not isinstance(arg, Variable):
+            raise TypeError('unknown type of  %s.' % type(arg))
+        
+        # Verify the memory_location of src
+        if arg.memory_location not in ('device', 'host', 'managed'):
+            raise ValueError("The direction of the copy should be from 'host' or 'device'")
+
+        # Verify the memory_location of dst
+        if memory_location not in ('device', 'host', 'managed'):
+            raise ValueError("The direction of the copy should be to 'host' or 'device'")
+        
+        # verify the type of is_async
+        if not isinstance(is_async, (LiteralTrue, LiteralFalse, bool)):
+            raise TypeError('is_async must be boolean')
+        
+        self._arg             = arg
+        self._shape           = arg.shape
+        self._rank            = arg.rank
+        self._dtype           = arg.dtype
+        self._order           = arg.order
+        self._precision       = arg.precision
+        self._memory_location = memory_location
+        self._is_async        = is_async
+        super().__init__()
+    
+    @property
+    def arg(self):
+        return self._arg
+
+    @property
+    def memory_location(self):
+        return self._memory_location
+
+    @property
+    def is_async(self):
+        return self._is_async
+
+class CudaThreadIdx(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaBlockDim(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaBlockIdx(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaGridDim(CudaInternalVar):
+    __slots__ = ()
+    pass
+
+class CudaGrid(PyccelAstNode)               :
+    """
+    CudaGrid locate Thread In the GPU architecture Using CudaThreadIdx, CudaBlockDim, CudaBlockIdx
+    To calculate the exact index of the thread automatically.
```

Comment:
Some of these docstrings have capitalisation in the middle of the sentence. This needs fixing it is not so clear to read. In English capital letters are used for the first word in a sentence, the name of something, or abbreviations.

```suggestion
    CudaGrid locates a thread in the GPU architecture using `CudaThreadIdx`, `CudaBlockDim`, and `CudaBlockIdx`
    to calculate the exact index of the thread automatically.
```